File: services/notifications/internal/domain/notification.go
// Package domain holds the GORM models and DTOs for the notifications service.
//
// v1.0 Notifications Engine — workstream notifications, Phase 1.
// Reference design doc: docs/superpowers/specs/2026-05-11-notifications-engine-design.md
package domain

import (
	"time"

	"gorm.io/datatypes"
)

// NotificationType enumerates the kinds of notifications the engine emits.
// v1.0 ships only "new_episode"; later phases add more types (e.g.
// "ongoing_resumed", "system_message") without altering the schema.
type NotificationType string

const (
	// TypeNewEpisode signals that one or more watched provider/team combos
	// detected new episodes for an anime. Detector output is grouped per
	// user + anime; the payload retains one representative combo for source
	// display while its episode range covers the grouped result.
	TypeNewEpisode NotificationType = "new_episode"

	// Feedback triage loop (AUTO-417): the player service emits one
	// notification per triage stage of a user-submitted feedback report.
	// Each stage invalidates the previous stage's unread notification for
	// the same report (via UpsertRequest.InvalidateDedupeKeys). Payload
	// shape for all three: FeedbackStatusPayload.

	// TypeFeedbackCreated — "we received your feedback and opened a task".
	TypeFeedbackCreated NotificationType = "feedback_created"
	// TypeFeedbackInProgress — "the robot started working on your task".
	TypeFeedbackInProgress NotificationType = "feedback_in_progress"
	// TypeFeedbackAIDone — "the robot finished, thanks for the feedback".
	TypeFeedbackAIDone NotificationType = "feedback_ai_done"
)

// UserNotification is the per-user notification row.
//
// Schema notes (design doc §Data Model):
//   - payload is JSONB so the same table can carry every NotificationType.
//   - read_at / dismissed_at / deleted_at / clicked_at are nullable timestamps
//     acting as state flags AND telemetry. dismissed_at ("cleared from the
//     bell, still visible in history") is distinct from deleted_at ("removed
//     from history too — the user hit the bin in the All-notifications modal").
//   - The two partial indexes GORM cannot express are created by
//     repo.EnsureIndexes after AutoMigrate:
//   - uk_user_dedupe UNIQUE (user_id, dedupe_key) WHERE dismissed_at IS NULL
//     AND deleted_at IS NULL (lets a fresh new_episode re-fire after the user
//     dismisses OR deletes the previous)
//   - idx_user_unread (user_id, created_at DESC) WHERE dismissed_at IS NULL
//     AND invalidated_at IS NULL AND deleted_at IS NULL (bell/dropdown path)
type UserNotification struct {
	ID            string         `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	UserID        string         `gorm:"type:uuid;not null;index" json:"user_id"`
	Type          string         `gorm:"size:32;not null;index" json:"type"`
	DedupeKey     string         `gorm:"size:255;not null" json:"dedupe_key"`
	Payload       datatypes.JSON `gorm:"type:jsonb;not null" json:"payload"`
	ReadAt        *time.Time     `json:"read_at"`
	DismissedAt   *time.Time     `gorm:"index" json:"dismissed_at"`
	InvalidatedAt *time.Time     `gorm:"index" json:"invalidated_at"`
	// DeletedAt is a plain lifecycle stamp (NOT gorm.DeletedAt — this service
	// manages every visibility filter explicitly, like dismissed_at, and must
	// never let GORM auto-scope Upsert/Get/cleanup). Set when the user deletes
	// a notification from the history modal; excluded from every list surface.
	DeletedAt *time.Time `gorm:"index" json:"deleted_at"`
	ClickedAt *time.Time `json:"clicked_at"`
	CreatedAt time.Time  `gorm:"index" json:"created_at"`
	UpdatedAt time.Time  `json:"updated_at"`
}

// TableName pins the table name so it does not depend on GORM's pluralization
// heuristic. (GORM would derive "user_notifications" either way; explicit
// override insulates the schema from future GORM changes.)
func (UserNotification) TableName() string { return "user_notifications" }

// FeedbackStatusPayload is the JSON shape stored in UserNotification.Payload
// for the three feedback_* types. ReportID is the on-disk report id
// (`{ts}_{user}_{type}` — see services/player report.go); Description is a
// truncated snippet of the user's original feedback text so the card can
// remind them which report this is about.
type FeedbackStatusPayload struct {
	ReportID    string `json:"report_id"`
	Category    string `json:"category,omitempty"` // bug | issue | feature
	Description string `json:"description,omitempty"`
	Status      string `json:"status"` // created | in_progress | ai_done
}

// NewEpisodePayload is the JSON shape stored in UserNotification.Payload
// when Type == TypeNewEpisode. Mirrors the design-doc payload spec.
// All fields lowercase_snake_case per the project's JSON convention.
type NewEpisodePayload struct {
	AnimeID                string `json:"anime_id"`
	ShikimoriID            string `json:"shikimori_id,omitempty"`
	AnimeTitle             string `json:"anime_title"`
	AnimePosterURL         string `json:"anime_poster_url,omitempty"`
	FirstUnwatchedEpisode  int    `json:"first_unwatched_episode"`
	LatestAvailableEpisode int    `json:"latest_available_episode"`
	Player                 string `json:"player"`
	Language               string `json:"language"`
	WatchType              string `json:"watch_type"`
	TranslationID          string `json:"translation_id"`
	TranslationTitle       string `json:"translation_title,omitempty"`
	WatchURL               string `json:"watch_url"`
}

File: frontend/web/src/types/notification.ts
/**
 * Notifications engine — v1.0 TypeScript types.
 *
 * Mirrors `services/notifications/internal/domain/notification.go` verbatim.
 * Snake_case field names match the backend JSON tags; do NOT camelCase them
 * here or the response envelope unwrap will silently drop fields.
 *
 * Workstream: notifications, Phase 3 (frontend bell + dropdown + toast).
 */

/**
 * NotificationType enumerates the kinds of notifications the engine emits.
 * v1.0 ships only `new_episode`; later phases add more types (e.g.
 * `ongoing_resumed`, `system_message`) without altering the schema.
 *
 * The store + renderer registry are payload-agnostic — adding a new type
 * is purely additive: ship a new renderer, define its payload type, no
 * changes to the bell/dropdown/toast.
 */
export type NotificationType =
  | 'new_episode'
  | 'feedback_created'
  | 'feedback_in_progress'
  | 'feedback_ai_done'
  | string

/**
 * Payload shape for `type === 'new_episode'`. Mirror of
 * `domain.NewEpisodePayload`. All fields lowercase_snake_case per the
 * project's JSON convention.
 *
 * `watch_url` ships as `/anime/{id}/watch?player=X&episode=N&translation=Y`
 * from the backend; the frontend route is `/anime/:id` with `?episode=N`
 * query — see `translateWatchUrl` in `stores/notifications.ts` and the
 * router alias in `router/index.ts`.
 */
export interface NewEpisodePayload {
  anime_id: string
  shikimori_id?: string
  anime_title: string
  anime_poster_url?: string
  first_unwatched_episode: number
  latest_available_episode: number
  player: string
  language: string
  watch_type: string
  translation_id: string
  translation_title?: string
  watch_url: string
}

/**
 * Payload shape for the three `feedback_*` types (AUTO-417 feedback triage
 * loop). Mirror of `domain.FeedbackStatusPayload` in the notifications
 * service. One card component renders all three stages; the stage itself is
 * carried both in `notification.type` and in `status`.
 */
export interface FeedbackStatusPayload {
  report_id: string
  category?: string // bug | issue | feature
  description?: string // truncated snippet of the user's original feedback
  status: string // created | in_progress | ai_done
}

/**
 * UserNotification — per-user notification row.
 *
 * `read_at` / `dismissed_at` / `deleted_at` / `clicked_at` are nullable
 * timestamps acting as both state flags and lightweight telemetry. The bell
 * badge counts rows where `read_at === null && dismissed_at === null`.
 * `dismissed_at` ("cleared from the bell, still shown in history") is distinct
 * from `deleted_at` ("binned from the All-notifications modal — gone from
 * history too").
 *
 * `payload` is typed `unknown` here because the engine carries every type
 * in the same table; downstream renderers cast to the concrete payload
 * shape (`as NewEpisodePayload`).
 */
export interface UserNotification {
  id: string
  user_id: string
  type: NotificationType
  dedupe_key: string
  payload: unknown
  read_at: string | null
  dismissed_at: string | null
  deleted_at: string | null
  clicked_at: string | null
  created_at: string
  updated_at: string
}

/**
 * Response shape for `GET /api/notifications?status=unread|all&limit=...`.
 * One round-trip surfaces the rows plus both counts the bell badge needs.
 */
export interface NotificationListResponse {
  notifications: UserNotification[]
  unread_count: number
  total: number
}

/** Response shape for `POST /api/notifications/mark-all-read`. */
export interface MarkAllReadResponse {
  updated: number
}
